rundk/main.go

```diff
@@ -0,0 +1,192 @@
+	mandatoryEnvVars          []string
+	optionalEnvVars           []string
+)
+
+func main() {
+	pflag.StringVar(&image, "test-image", "gcr.io/knative-tests/test-infra/prow-tests:stable", "The image we use to run the test flow.")
+	pflag.StringVar(&entrypoint, "entrypoint", "runner.sh", "The entrypoint executable that runs the test commands.")
+	pflag.BoolVar(&dindEnabled, "enable-docker-in-docker", false, "Enable running docker commands in the test flow. "+
+		"By enabling this the container will share the same docker daemon in the host machine, so be careful when using it.")
+	pflag.BoolVar(&useLocalGcloudCredentials, "use-local-gcloud-credentials", false, "Use the same gcloud credentials as local, which can be set "+
+		"either by setting env var GOOGLE_CLOUD_APPLICATION_CREDENTIALS or from ~/.config/gcloud")
+	pflag.BoolVar(&useLocalKubeconfig, "use-local-kubeconfig", false, "Use the same kubeconfig as local, which can be set "+
+		"either by setting env var KUBECONFIG or from ~/.kube/config")
+	pflag.StringSliceVar(&mounts, "mounts", []string{}, "A list of extra folders or files separated by comma that need to be mounted to run the test flow."+
+		"It must be in the format of `source1:target1,source2:target2,source3:target3`.")
+	pflag.StringSliceVar(&mandatoryEnvVars, "mandatory-env-vars", []string{}, "A list of env vars separated by comma that must be set on local.")
+	pflag.StringSliceVar(&optionalEnvVars, "optional-env-vars", []string{}, "A list of env vars separated by comma that optionally need to be set on local.")
+	pflag.Parse()
+
+	cmd, cancel := setup()
+	defer cancel()
+
+	run(cmd, pflag.Args()...)
+}
+
+func setup() (interactive.Docker, func()) {
+	var err error
+
+	// Until everyone is using 20.xx version of docker (see https://github.com/docker/cli/pull/1498) adding the --pull flag,
+	//  need to separately pull the image first to be sure we have the latest
+	pull := interactive.NewCommand("docker", "pull", image)
+	if err = pull.Run(); err != nil {
+		log.Fatal("Error pulling the test image: ", err)
+	}
+
+	builtUpDefers := make([]func(), 0)
+
+	// Setup command
+	cmd := interactive.NewDocker()
+	envs := interactive.Env{}
+
+	// Setup temporary directory to save the artifacts
+	tmpDir, err := ioutil.TempDir("/tmp", "prow-docker.")
+	if err != nil {
+		log.Fatal("Error setting up the temporary directory: ", err)
+	}
+
+	// Setup docker-in-docker
+	if dindEnabled {
+		cmd.AddMount("bind", "/var/run/docker.sock", "/var/run/docker.sock")
+		// On a MAC machine, set to host networking to make DinD work. Not
+		// required on Linux.
+		if runtime.GOOS == "darwin" {
+			cmd.AddArgs("--network", "host")
+		}
+	}
+
+	// Setup gcloud credentials
+	if useLocalGcloudCredentials {
+		gcloudKey := os.Getenv("GOOGLE_APPLICATION_CREDENTIALS")
+		// If GOOGLE_APPLICATION_CREDENTIALS is not empty, also mount the key file
+		// to the container
+		if gcloudKey != "" {
+			envs["GOOGLE_APPLICATION_CREDENTIALS"] = gcloudKey
```

Comment:
You should use a constant filename for the file inside the image

Reply:
Hmm do you mean always mounting the key file to the same target path? If yes are there any reasons or advantages?

Reply:
Yes:
Privacy: It leaks information about the caller. (it's mounting their home directory setup, not the biggest deal in the world)
Debugging: Easier to know the file will be in a fixed position (Chi: running this doesn't work for me; Cory: can you `cat /gac.json`?)
Defense-in-depth: Weird file paths could break things in scripts, setting to a fixed, simple name avoids that

Reply:
1. I don't think it will leak information about the caller, they will run the docker command and start the container on their local anyway..
2. If `GOOGLE_APPLICATION_CREDENTIALS` is set on the local, it must be explicitly set by the user, so they must know where the file is, and I don't think it's harder to find than asking us the magic value we hardcoded.
3. I don't seem to realize there could be weird file paths that could break things, especially since we only mount it as a single file. Do you have a concrete example or have you met issues related to that?